File: clients/java-sdk/src/main/java/io/flowcatalyst/sdk/tsid/Tsid.java
package io.flowcatalyst.sdk.tsid;

import java.util.concurrent.ThreadLocalRandom;

/**
 * Lightweight TSID (Time-Sorted ID) generator, byte-for-byte compatible with
 * the TypeScript SDK: 13-character Crockford Base32 strings from a 64-bit
 * value composed of 42 bits of timestamp (custom epoch 2020-01-01T00:00:00Z)
 * + 22 bits of randomness.
 */
public final class Tsid {

    private static final String CROCKFORD_ALPHABET = "0123456789ABCDEFGHJKMNPQRSTVWXYZ";
    private static final int TSID_LENGTH = 13;
    private static final int RANDOM_BITS = 22;
    private static final int RANDOM_MASK = (1 << RANDOM_BITS) - 1;
    private static final long CUSTOM_EPOCH = 1577836800000L;

    /**
     * Packed (millis, sequence) state for collision-free generation: the
     * sequence is randomly seeded each new millisecond and incremented within
     * it (borrowing into the next millisecond on overflow) — same scheme as
     * the platform's Go generator, same wire layout as the TypeScript SDK.
     */
    private static final java.util.concurrent.atomic.AtomicLong STATE = new java.util.concurrent.atomic.AtomicLong(0);

    private Tsid() {}

    /** Generate a new TSID as a 13-character Crockford Base32 string. */
    public static String generate() {
        while (true) {
            long previous = STATE.get();
            long previousMs = previous >>> RANDOM_BITS;
            long nowMs = System.currentTimeMillis() - CUSTOM_EPOCH;

            long next;
            if (nowMs > previousMs) {
                next = (nowMs << RANDOM_BITS)
                        | ThreadLocalRandom.current().nextInt(RANDOM_MASK + 1);
            } else {
                // Same (or clock-rewound) millisecond: increment; on sequence
                // wrap, borrow into the next millisecond.
                next = previous + 1;
            }
            if (STATE.compareAndSet(previous, next)) {
                return encodeCrockford(next);
            }
        }
    }

    /**
     * Generate a BRANDED (typed) TSID: {@code {prefix}_{raw}} — matching the
     * FlowCatalyst platform convention (e.g. {@code aud_…}, {@code prn_…}).
     * Use a short lowercase prefix for your own entities, e.g.
     * {@code generateWithPrefix("cmt")} → {@code cmt_6F7JC2A6JFR7N}.
     *
     * @throws IllegalArgumentException if the prefix is empty or contains an underscore
     */
    public static String generateWithPrefix(String prefix) {
        if (prefix == null || prefix.isEmpty() || prefix.contains("_")) {
            throw new IllegalArgumentException(
                    "TSID prefix must be non-empty and contain no underscore.");
        }
        return prefix + "_" + generate();
    }

    /** Validate that a string is a valid (unbranded) TSID format. */
    public static boolean isValid(String tsid) {
        if (tsid == null || tsid.length() != TSID_LENGTH) return false;
        String upper = tsid.toUpperCase(java.util.Locale.ROOT);
        for (int i = 0; i < upper.length(); i++) {
            if (CROCKFORD_ALPHABET.indexOf(upper.charAt(i)) < 0) return false;
        }
        return true;
    }

    private static String encodeCrockford(long value) {
        char[] chars = new char[TSID_LENGTH];
        long remaining = value;
        for (int i = TSID_LENGTH - 1; i >= 0; i--) {
            chars[i] = CROCKFORD_ALPHABET.charAt((int) (remaining & 31L));
            remaining >>>= 5;
        }
        return new String(chars);
    }
}

File: clients/typescript-sdk/src/outbox/tsid.ts
/**
 * Lightweight TSID (Time-Sorted ID) generator.
 *
 * Generates 13-character Crockford Base32 strings from a 64-bit value:
 * 42 bits of timestamp, then 22 low bits split as Go's `pkg/fcsdk/tsid`
 * does — a 12-bit sequence above 10 random bits. The sequence starts at a
 * random value each millisecond and increments within it, so ids from one
 * process are strictly increasing: outbox rows written in one transaction
 * (one `created_at`) keep their creation order when sorted by id, which the
 * outbox processor relies on within a message group.
 */

const CROCKFORD_ALPHABET = "0123456789ABCDEFGHJKMNPQRSTVWXYZ";
const TSID_LENGTH = 13;
const LOW_BITS = 22;
const SEQUENCE_MAX = 0xfff;
const RANDOM_MASK = 0x3ff;

// Custom epoch: 2020-01-01T00:00:00Z
const CUSTOM_EPOCH = 1577836800000;

let lastMs = -1;
let lastSeq = 0;

function randomBelow(n: number): number {
	return Math.floor(Math.random() * n);
}

/** The next (millisecond, sequence) pair (Go `nextMsSeq`). */
function nextMsSeq(): [number, number] {
	const now = Date.now() - CUSTOM_EPOCH;
	if (now > lastMs) {
		lastMs = now;
		lastSeq = randomBelow(SEQUENCE_MAX + 1);
	} else if (lastSeq < SEQUENCE_MAX) {
		lastSeq += 1;
	} else {
		lastMs += 1;
		lastSeq = randomBelow(SEQUENCE_MAX + 1);
	}
	return [lastMs, lastSeq];
}

/**
 * Generate a new TSID as a 13-character Crockford Base32 string.
 */
export function generate(): string {
	const [ms, seq] = nextMsSeq();
	const random = randomBelow(RANDOM_MASK + 1);
	// Use BigInt for 64-bit arithmetic
	const value =
		(BigInt(ms) << BigInt(LOW_BITS)) | (BigInt(seq) << 10n) | BigInt(random);
	return encodeCrockford(value);
}

/**
 * Generate a BRANDED (typed) TSID: `${prefix}_${raw}` — matching the
 * FlowCatalyst platform convention (e.g. `aud_…`, `prn_…`). Use a short
 * lowercase prefix for your own entities, e.g. `generateWithPrefix("cmt")`
 * → `cmt_6F7JC2A6JFR7N`.
 *
 * @throws if the prefix is empty or contains an underscore.
 */
export function generateWithPrefix(prefix: string): string {
	if (prefix.length === 0 || prefix.includes("_")) {
		throw new Error("TSID prefix must be non-empty and contain no underscore.");
	}
	return `${prefix}_${generate()}`;
}

/**
 * Validate that a string is a valid TSID format.
 */
export function isValid(tsid: string): boolean {
	if (tsid.length !== TSID_LENGTH) return false;
	const upper = tsid.toUpperCase();
	for (let i = 0; i < upper.length; i++) {
		if (!CROCKFORD_ALPHABET.includes(upper[i]!)) return false;
	}
	return true;
}

function encodeCrockford(value: bigint): string {
	const chars: string[] = Array.from({ length: TSID_LENGTH });
	let remaining = value;

	for (let i = TSID_LENGTH - 1; i >= 0; i--) {
		chars[i] = CROCKFORD_ALPHABET[Number(remaining & 31n)]!;
		remaining >>= 5n;
	}

	return chars.join("");
}
